pkg/util/kubernetes/hpa/hpa.go

```diff
@@ -0,0 +1,387 @@
+	namespace := as.GetResourcesNamespace()
+	clientAPI, err := as.GetAPIClient()
+	datadogCl, err := NewDatadogClient()
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
+		log.Errorf("Error creating Client for the HPA: %s", err.Error())
+		return nil, err
+	}
+	pollInterval := config.Datadog.GetInt("hpa_watcher_polling_freq")
+	refreshInterval := config.Datadog.GetInt("hpa_external_metrics_polling_freq")
+	externalMaxAge := config.Datadog.GetInt("hpa_external_metrics_max_age")
+	return &HPAWatcherClient{
+		clientSet:      clientAPI.Cl,
+		readTimeout:    100 * time.Millisecond,
+		pollItl:        time.NewTicker(time.Duration(pollInterval) * time.Second),
+		refreshItl:     time.NewTicker(time.Duration(refreshInterval) * time.Second),
+		externalMaxAge: time.Duration(externalMaxAge) * time.Second,
+		ns:             namespace,
+		datadogClient:  datadogCl,
+	}, nil
+}
+
+// Start runs a watch process of the various HPA objects' activities to process and store the relevant info.
+// Refreshes the custom metrics stored as well.
+func (c *HPAWatcherClient) Start() {
+	log.Info("Starting HPA Process ...")
+	tickerHPAWatchProcess := c.pollItl
+	tickerHPARefreshProcess := c.refreshItl
+
+	var resversion string
+	err := c.createConfigMapHPA()
+	if err != nil {
+		log.Errorf("Could not retrieve the ConfigMap %s to run the HPA process, stopping: %s", datadogHPAConfigMap, err.Error())
+		return
+	}
+
+	// Creating a leader election engine to make sure only the leader writes the metrics in the configmap and queries Datadog.
+	leaderEngine, err := leaderelection.GetLeaderEngine()
+	if err != nil {
+		log.Errorf("Could not ensure the leader election is running properly: %s", err)
+		return
+	}
+	leaderEngine.EnsureLeaderElectionRuns()
+
+	go func() {
+		for {
+			select {
+			// Ticker for the HPA Object watcher
+			case <-tickerHPAWatchProcess.C:
+				if !leaderEngine.IsLeader() {
+					continue
+				}
+				newHPA, modified, deleted, res, err := c.run(resversion)
+				if err != nil {
+					log.Errorf("Error while watching HPA Objects' activities: %s", err)
+					return
+				}
+				if res != resversion && res != "" {
+					resversion = res
+					if len(newHPA) > 0 {
+						c.storeHPA(newHPA)
+					}
+					if len(modified) > 0 {
+						c.storeHPA(modified)
+					}
+					if len(deleted) > 0 {
+						log.Infof("deleting if resver is diff")
+						c.removeEntryFromConfigMap(deleted)
+					}
+				}
+			// Ticker to run the refresh process for the stored external metrics
+			case <-tickerHPARefreshProcess.C:
+				if !leaderEngine.IsLeader() {
+					continue
+				}
+				// Updating the metrics against Datadog should not affect the HPA pipeline.
+				// If metrics are temporarily unavailable for too long, they will become `Valid=false` and won't be evaluated.
+				c.updateCustomMetrics()
+			}
+		}
+	}()
+}
+
+func (c *HPAWatcherClient) updateCustomMetrics() {
+	maxAge := int64(c.externalMaxAge.Seconds())
+	var key string
```

Comment:
not sure was this is here, you could just `key :=` below ?

Reply:
I thought it was more efficient memory wise to allocate the variable before the for loop.